### PR TITLE
added RateLimiter.drainPermissions method

### PR DESCRIPTION
**Pull request by walec51**
First part of https://github.com/resilience4j/resilience4j/issues/1228

### CODE REVIEW

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/RateLimiterOnDrainEvent.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+public class RateLimiterOnDrainEvent extends AbstractRateLimiterEvent {
+
+    public RateLimiterOnDrainEvent(String rateLimiterName, int numberOfPermits) {
+        super(rateLimiterName, numberOfPermits);
+    }
+
+    @Override
+    public Type getEventType() {
+        return Type.SUCCESSFUL_ACQUIRE;
```
Comment (RobWin) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/RateLimiterOnDrainEvent.java:29]: DRAINED

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/RateLimiterOnDrainEvent.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.ratelimiter.event;
+
+public class RateLimiterOnDrainEvent extends AbstractRateLimiterEvent {
```
Comment (RobWin) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/RateLimiterOnDrainEvent.java:21]: RateLimiterOnDrainedEvent 
wdyt?

**Review by RobWin — CHANGES_REQUESTED** [review]

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/RateLimiterOnDrainEvent.java (hunk 1 above)
Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/event/RateLimiterOnDrainEvent.java]: fixed

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java [hunk 3]
```diff
@@ -128,20 +129,34 @@ public long reservePermission(final int permits) {
...
             return modifiedState.nanosToWait;
         }
 
-        publishRateLimiterEvent(false, permits);
+        publishRateLimiterAcquisitionEvent(false, permits);
         return -1;
     }
 
+    @Override
+    public void drainPermissions() {
```
Comment (storozhukBM) [inline resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java:147]: Please don't change state directly, but use the same approach as in `updateStateWithBackOff` https://github.com/resilience4j/resilience4j/blob/a2534c515eeb9361b755bbaab2210990f99ac123/resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java#L157

**Review by storozhukBM — CHANGES_REQUESTED** [review]

File: resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java (hunk 3 above)
Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java:147]: I don't think that is a good idea

to drain the number of permits to zero precisely in a thread safe manner I have to read the number of permits left and subtract that amount in one atomic operation

`updateStateWithBackOff` doesn't do that - explain why using it would be better?

Comment (storozhukBM) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java:147]: `updateStateWithBackOff` uses `compareAndSwap` operation, so it is atomic, consistent and will do completely fine.
`getAndUpdate` is also fine with one exception that it will reduce throughput of contended operations on `state` for Java 8 (starting from Java 9 it is fine), for additional details, please read the doc comment on `updateStateWithBackOff` method.

Comment (walec51) [reply resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java:147]: reimplemented drain using compareAndSet

please check if this impl is ok for you

I also added a unit test for an drain edge case

**Comment by walec51** [comment]
@RobWin I think you broke the master branch https://travis-ci.org/github/resilience4j/resilience4j/builds/743885837

and now my branch is broken as well after resolving a commit

**Comment by RobWin** [comment]
Should be fixed now

**Comment by walec51** [comment]
@RobWin @storozhukBM can we merge this now?

**Comment by RobWin** [comment]
I will do it when I'm back from quarantine

**Review by RobWin — APPROVED** [review]
